consensus/src/liveness/cached_proposer_election.rs

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::BTreeMap;
+
+use aptos_infallible::Mutex;
+use consensus_types::common::{Author, Round};
+
+use super::proposer_election::ProposerElection;
+
+// Wrapper around ProposerElection.
+//
+// Function get_valid_proposer can be expensive, and we want to make sure
+// it is computed only once for a given round.
+// Additionally, provides is_valid_proposal that remembers, and rejects if
+// the same leader proposes multiple blocks.
+pub struct CachedProposerElection {
+    proposer_election: Box<dyn ProposerElection + Send + Sync>,
+    recent_elections: Mutex<BTreeMap<Round, Author>>,
+    window: Round,
+}
+
+impl CachedProposerElection {
+    pub fn new(proposer_election: Box<dyn ProposerElection + Send + Sync>, window: Round) -> Self {
+        Self {
+            proposer_election,
+            recent_elections: Mutex::new(BTreeMap::new()),
+            window,
+        }
+    }
+}
+
+impl ProposerElection for CachedProposerElection {
+    fn get_valid_proposer(&self, round: Round) -> Author {
+        let mut recent_elections = self.recent_elections.lock();
+
+        // Once map_first_last stabilized in Rust change:
+        //   while recent_elections.iter().next().map_or(false, |(key, _)| key + self.window < round) {
+        //     recent_elections.pop_first();
+        //   }
+        loop {
+            let entry = recent_elections.iter().next();
+            if entry.map_or(false, |(key, _)| key + self.window < round) {
```

Comment:
Hmm..this looks a bit confusing to me. Why not just use an LRU cache of a fixed size instead of using a `BTreeMap` as the cache here? That way we don't have to manually deal with cache evection and all.

Reply:
@sitalkedia - we want a sliding window cache - we know when we move to the next round, what is the oldest value we care about.

with LRU cache you can have a value with larger round be evicted before a value with lower round is, based on access pattern.

for example, if we ask for value for round 5, then for round 4 then for round 3, then for round 6, we are going to evict round 5, instead of round 3.

I'll add comment into the code on why we use BTreeMap

Reply:
@ikabiljo - Thanks for the explanation - yes, LRU cache might not be as accurate as sliding window cache in this case but I would say for a large enough window size that will be good enough. It just makes the code a bit simpler but I don't have a strong opinion about it.